### DISCUSSION TITLE
Heat IDs in V4 differ from V3?

### DISCUSSION THREAD

**Discussion by hurcane** [wixtoolset/issues#8145] [Questions] [answered]
We recently upgraded our project to use Wix V4 from V3. We have a few manual steps in our process where we run Heat from the command line. After doing the necessary setup to enable Heat from the command line, we are finding that the generated component and file IDs are different from what V3 produced. This is going to cause problems with upgrades...

This is one of the commands that we use:
`heat.exe dir .\Dashboard -ag -cg Dashboard -dr Dashboard -scom -sfrag -srd -sreg -var var.DashboardSource -o db.wxs`

V3 generates this example component:
```
<Component Id="cmp12D55E136F0665240BA4DAF0A6161673" Guid="*">
    <File Id="fil2DE5CF1980F16A2DD5E64092E109FBCC" KeyPath="yes" Source="$(var.DashboardSource)\Accessibility.dll" />
</Component>
```
V4 generates this:
```
<Component Id="cmpGlvmtLzdXBn8lx_jXDsieirLskA" Guid="*">
    <File Id="filgPc2_PANfu9vaVHgRXATCs8bAag" KeyPath="yes" Source="$(var.DashboardSource)\Accessibility.dll" />
</Component>
```

Do we need to use different parameters to ensure we get consistent results for our upgrades? Or do we need to install Wix V3 and use the V3 version of Heat?

**Comment by robmen** [comment]
> This is going to cause problems with upgrades...

Really? Why? Those are internal identifiers that only impact patching (which is one of many reasons patching with different versions of the WiX Toolset, even minor version differences, is not supported).

**Reply by hurcane** [reply]
Maybe I have a long-held misunderstanding of Wix, but don't major upgrades depend on the component's Id property staying consistent for files that are contained in both versions? We follow the model where removal of the previous version happens after the new version is installed.

**Reply by robmen** [reply, marked as answer]
The `Guid` needs to be stable. There can be some confusion for those that know the MSI table schema because the column that contains the Component's GUID is named `ComponentId`.

**Reply by hurcane** [reply]
Ah. yes, now it is coming back to me. As long as the install path and file name stay the same, the auto-generated GUID will be stable. So, other than an unusually large diff in the Git commit, the installer process should not be affected.
